storage/src/mmr/operation_array.rs

```diff
@@ -0,0 +1,220 @@
+//! An `Array` implementation for operations applied to a `MutableMmr` K/V store, making them
+//! storable in a `fixed::Journal`.
+
+use commonware_codec::{Codec, Error as CodecError, Reader, SizedCodec, Writer};
+use commonware_utils::Array;
+use std::{
+    cmp::{Ord, PartialOrd},
+    fmt::{Debug, Display},
+    hash::Hash,
+    ops::Deref,
+};
+use thiserror::Error;
+
+/// Errors returned by `Operation` functions.
+#[derive(Error, Debug)]
+pub enum Error<K: Array, V: Array> {
+    #[error("invalid length")]
+    InvalidLength,
+    #[error("invalid key: {0}")]
+    InvalidKey(<K as Array>::Error),
+    #[error("invalid value: {0}")]
+    InvalidValue(<V as Array>::Error),
+}
+
+/// The types of operations that change a key's state in the store.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub enum Type<V: Array> {
+    /// Indicates the key no longer has a value.
+    Deleted,
+
+    /// Indicates the key now has the wrapped value.
+    Update(V),
+}
+
+/// An `Array` implementation for operations applied to a `MutableMmr` K/V store.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+#[repr(transparent)]
+pub struct Operation<K: Array, V: Array> {
+    pub data: Vec<u8>,
+    _phantom: std::marker::PhantomData<(K, V)>,
+}
+
+impl<K: Array, V: Array> Operation<K, V> {
+    const DELETE_CONTEXT: u8 = 0;
+    const UPDATE_CONTEXT: u8 = 1;
+
+    /// Create a new operation from `key` and `op_type`.
+    pub fn new(key: K, op_type: Type<V>) -> Self {
+        let mut data = Vec::with_capacity(Self::LEN_ENCODED);
+        data.extend_from_slice(&key);
+        match op_type {
+            Type::Deleted => {
+                data.push(Self::DELETE_CONTEXT);
+                data.extend(vec![0u8; V::LEN_ENCODED]);
+            }
+            Type::Update(value) => {
+                data.push(Self::UPDATE_CONTEXT);
+                data.extend_from_slice(&value);
+            }
+        }
+
+        Self {
+            data,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+
+    pub fn to_key(&self) -> K {
+        K::try_from(&self.data[..K::LEN_ENCODED]).unwrap()
```

Comment:
I wonder if it makes sense to keep `key` around to prevent re-parsing like this?

Reply:
Parsing should be mostly just memcpy right?  I think this is mostly a transient type instantiated to "wrap" on-disk representation (the raw byte array).